feat: second mapping tutorial (#4720)
JetBrains/kotlin-web-site: docs/topics/native/mapping-struct-union-types-from-c.md

>
{style="warning"}

In this tutorial, you'll learn how C struct and union declarations are visible from Kotlin and explore the advanced C
interop-related usages of Kotlin/Native and [multiplatform](gradle-configure-project.md#targeting-multiple-platforms)
Gradle builds.

> In the tutorial, you'll learn:

Author: I agree that it's better to avoid this repetition, I suggested a different version

> In the tutorial, you'll learn:

Reviewer: I like it 👍

        // macosX64("native") {   // macOS on x86_64 platforms
        // linuxArm64("native") { // Linux on ARM64 platforms
        // linuxX64("native") {   // Linux on x86_64 platforms
        // mingwX64("native") {   // Windows
            compilations.main.cinterops {
                interop {   
                    definitionFile = project.file('src/nativeInterop/cinterop/interop.def')
                }
            }
        
            binaries {
                executable()
            }
        }
    }
    ```

    </tab>
    </tabs> 
   
3. With the help of IntelliJ IDEA's [Go to declaration](https://www.jetbrains.com/help/rider/Navigation_and_Search__Go_to_Declaration.html)
   command (<shortcut>Cmd + B</shortcut>/<shortcut>Ctrl + B</shortcut>), you can navigate to the following generated API
   for C functions, struct, and union:

   ```kotlin
   fun struct_by_value(s: kotlinx.cinterop.CValue<interop.MyStruct>)
   fun struct_by_pointer(s: kotlinx.cinterop.CValuesRef<interop.MyStruct>?)
   
   fun union_by_value(u: kotlinx.cinterop.CValue<interop.MyUnion>)
   fun union_by_pointer(u: kotlinx.cinterop.CValuesRef<interop.MyUnion>?)
   ```

> You see that `cinterop` generates Kotlin types for struct and union declarations in C. The generated API includes

Reviewer: nice, I like the new version 👍

object instance. The function requires a [lambda function with a receiver](lambdas.md#function-literals-with-receiver) 
to initialize the underlying C type in-place. The function is declared as follows:

```kotlin
fun <reified T : CStructVar> cValue(initialize: T.() -> Unit): CValue<T>
```

Now it's time to see how to use `cValue` and pass by-value parameters:

```kotlin
import interop.*
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.cValue

@OptIn(ExperimentalForeignApi::class)
fun callValue() {

    val cStruct = cValue<MyStruct> {
        a = 42
        b = 3.14
    }
    struct_by_value(cStruct)

    val cUnion = cValue<MyUnion> {
        b.a = 5
        b.b = 2.7182
    }

    union_by_value(cUnion)
}
```

### Create struct and union as CValuesRef&lt;T&gt;

The `CValuesRef<T>` type is used in Kotlin to pass a typed pointer parameter of a C function. To allocate `MyStruct` and
`MyUnion` in the native memory, use the following extension function on `kotlinx.cinterop.NativePlacement` type:

```kotlin
fun <reified T : kotlinx.cinterop.CVariable> alloc(): T
```

> `NativePlacement` represents native memory with functions similar to `malloc` and `free`. There are several implementations

Author: I'd refrain from making changes here, it creates a different meaning
So the idea is "`NativePlacement` has several implementations, global one is `kotlinx.cinterop.nativeHeap`, `memScoped()` is an alternative

> `NativePlacement` represents native memory with functions similar to `malloc` and `free`. There are several implementations

Reviewer: ah I see what you mean! - looks good then I got confused by `malloc` and `free` vs `nativeHeap.free`

of `NativePlacement`:

* The global implementation is `kotlinx.cinterop.nativeHeap`, but you must call `nativeHeap.free()` to release the memory
  after use.
* A safer alternative is `memScoped()`, which creates a short-lived memory scope where all allocations are automatically
  freed at the end of the block:

  ```kotlin
  fun <R> memScoped(block: kotlinx.cinterop.MemScope.() -> R): R
  ```

> With `memScoped()`, your code to call functions with pointers will look like this:

Reviewer: ```suggestion
With `memScoped()`, your code to call functions with pointers looks like this:
```

```kotlin
import interop.*
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.memScoped
import kotlinx.cinterop.alloc
import kotlinx.cinterop.ptr

@OptIn(ExperimentalForeignApi::class)
fun callRef() {
    memScoped {
        val cStruct = alloc<MyStruct>()
        cStruct.a = 42
        cStruct.b = 3.14

        struct_by_pointer(cStruct.ptr)

        val cUnion = alloc<MyUnion>()
        cUnion.b.a = 5
        cUnion.b.b = 2.7182

        union_by_pointer(cUnion.ptr)
    }
}
```

Here, the `ptr` extension property, which is available within the `memScoped {}` block, converts `MyStruct` and `MyUnion`
instances into native pointers.

> Since memory is managed inside the `memScoped {}` block, it will be automatically freed at the end of the block. Ensure

Reviewer: ```suggestion
Since memory is managed inside the `memScoped {}` block, it is automatically freed at the end of the block. Ensure
```

> not to use pointers outside of this scope to avoid accessing deallocated memory. If you need longer-lived allocations

Reviewer: (suggestion)
What do you think of this version? 

Avoid using pointers outside of this scope to prevent accessing deallocated memory.

(for example, for caching in a C library), consider using `Arena()` or `nativeHeap`.

### Conversion between CValue&lt;T&gt; and CValuesRef&lt;T&gt;

> There are cases when you need to pass a struct as a value to one call and then pass the same struct

Reviewer: ```suggestion
Sometimes, you need to pass a struct as a value in one function call and then pass the same struct
```

> as a reference to another call.

Reviewer: ```suggestion
as a reference in another.
```

> To do that, you'll need a `NativePlacement`, but first, let's see how `CValue<T>` is turned to a pointer:

Reviewer: ```suggestion
To do this, you need a `NativePlacement`, but first, let's see how `CValue<T>` is turned into a pointer:
```

```kotlin
import interop.*
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.cValue
import kotlinx.cinterop.memScoped

@OptIn(ExperimentalForeignApi::class)
fun callMix_ref() {
    val cStruct = cValue<MyStruct> {
        a = 42
        b = 3.14
    }

    memScoped { 
        struct_by_pointer(cStruct.ptr)
    }
}
```

Here again, the `ptr` extension property from `memScoped {}` turns `MyStruct` instances into native pointers.
These pointers are only valid inside the `memScoped {}` block.

> For the opposite conversion, to turn a pointer into a by-value variable, call the `readValue()` extension function:

Reviewer: ```suggestion
To convert a pointer back into a by-value variable, call the `.readValue()` extension function:
```

```kotlin
import interop.*
import kotlinx.cinterop.alloc
import kotlinx.cinterop.ExperimentalForeignApi
import kotlinx.cinterop.memScoped
import kotlinx.cinterop.readValue

@OptIn(ExperimentalForeignApi::class)
fun callMix_value() {
    memScoped {
        val cStruct = alloc<MyStruct>()
        cStruct.a = 42
        cStruct.b = 3.14

        struct_by_value(cStruct.readValue())
    }
}
```

## Update Kotlin code

Now that you've learned how to use C declarations in Kotlin code, try to use them in your project.

> The final code in the `hello.kt` file may look like this:

Reviewer: ```suggestion
Your `hello.kt` file should now look like this:
```

Author: The problem here is that the previous steps don't contribute to updating the main function directly, it's just a learning material on how users may change the code in `main()`. so I kept it as is
